mathics/builtin/forms.py

```diff
@@ -0,0 +1,520 @@
+
+    options = {"TableDepth": "Infinity"}
+    summary_text = "format as a table"
+
+    def apply_makeboxes(self, table, f, evaluation, options):
+        """MakeBoxes[%(name)s[table_, OptionsPattern[%(name)s]],
+        f:StandardForm|TraditionalForm|OutputForm]"""
+
+        dims = len(get_dimensions(table, head=SymbolList))
+        depth = self.get_option(options, "TableDepth", evaluation)
+        depth = expr_min((Integer(dims), depth))
+        depth = depth.value
+        if depth is None:
+            evaluation.message(self.get_name(), "int")
+            return
+
+        if depth <= 0:
+            return Expression(SymbolMakeBoxes, table, f)
+        elif depth == 1:
+            return GridBox(
+                ListExpression(
+                    *(
+                        ListExpression(Expression(SymbolMakeBoxes, item, f))
+                        for item in table.elements
+                    ),
+                )
+            )
+            # return Expression(
+            #    'GridBox', Expression('List', *(
+            #        Expression('List', Expression('MakeBoxes', item, f))
+            #        for item in table.elements)))
+        else:
+            new_depth = Expression(SymbolRule, SymbolTableDepth, Integer(depth - 2))
+
+            def transform_item(item):
+                if depth > 2:
+                    return Expression(Symbol(self.get_name()), item, new_depth)
+                else:
+                    return item
+
+            return GridBox(
+                ListExpression(
+                    *(
+                        ListExpression(
+                            *(
+                                Expression(SymbolMakeBoxes, transform_item(item), f)
+                                for item in row.elements
+                            ),
+                        )
+                        for row in table.elements
+                    ),
+                )
+            )
+
+
+class MatrixForm(TableForm):
+    """
+    <dl>
+      <dt>'MatrixForm[$m$]'
+      <dd>displays a matrix $m$, hiding the underlying list
+        structure.
+    </dl>
+
+    >> Array[a,{4,3}]//MatrixForm
+     = a[1, 1]   a[1, 2]   a[1, 3]
+     .
+     . a[2, 1]   a[2, 2]   a[2, 3]
+     .
+     . a[3, 1]   a[3, 2]   a[3, 3]
+     .
+     . a[4, 1]   a[4, 2]   a[4, 3]
+
+    ## Issue #182
+    #> {{2*a, 0},{0,0}}//MatrixForm
+     = 2 a   0
+     .
+     . 0     0
+    """
+
+    summary_text = "format as a matrix"
+
+    def apply_makeboxes_matrix(self, table, f, evaluation, options):
+        """MakeBoxes[%(name)s[table_, OptionsPattern[%(name)s]],
+        f:StandardForm|TraditionalForm]"""
+
+        result = super(MatrixForm, self).apply_makeboxes(table, f, evaluation, options)
+        if result.get_head_name() == "System`GridBox":
+            return Expression(
+                SymbolRowBox, ListExpression(String("("), result, String(")"))
+            )
+        return result
+
+
+class StringForm(Builtin):
```

Comment:
On new or revised code, let's please put classes in alphabetic order unless there is a need not to. Thanks.